src/Common/Scheduler/CPULeaseAllocation.cpp

```diff
@@ -0,0 +1,533 @@
+    extern const Event ConcurrencyControlUpscales;
+    extern const Event ConcurrencyControlDownscales;
+}
+
+namespace CurrentMetrics
+{
+    extern const Metric ConcurrencyControlScheduled;
+    extern const Metric ConcurrencyControlAcquired;
+    extern const Metric ConcurrencyControlPreempted;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int RESOURCE_ACCESS_DENIED;
+}
+
+std::atomic<size_t> CPULeaseAllocation::lease_counter{0};
+
+CPULeaseAllocation::Lease::Lease(CPULeaseAllocationPtr && parent_, size_t slot_id_)
+    : ISlotLease(slot_id_)
+    , parent(std::move(parent_))
+{}
+
+CPULeaseAllocation::Lease::~Lease()
+{
+    if (parent)
+    {
+        std::optional<OpenTelemetry::SpanHolder> span;
+        if (parent->settings.trace_cpu_scheduling)
+        {
+            span.emplace("CPU_LEASE_STOP");
+            span->addAttribute("workload", parent->settings.workload);
+            span->addAttribute("lease_id", parent->getLeaseId());
+            span->addAttribute("thread_number", slot_id);
+        }
+        parent->release(*this);
+    }
+}
+
+void CPULeaseAllocation::Lease::startConsumption()
+{
+    last_report_ns = clock_gettime_ns(CLOCK_THREAD_CPUTIME_ID);
+    if (parent && parent->settings.trace_cpu_scheduling)
+    {
+        OpenTelemetry::SpanHolder span("CPU_LEASE_START");
+        span.addAttribute("workload", parent->settings.workload);
+        span.addAttribute("lease_id", parent->getLeaseId());
+        span.addAttribute("thread_number", slot_id);
+    }
+}
+
+bool CPULeaseAllocation::Lease::renew()
+{
+    if (parent)
+        return parent->renew(*this);
+    else
+        return false;
+}
+
+void CPULeaseAllocation::Lease::reset()
+{
+    if (parent->settings.trace_cpu_scheduling)
+    {
+        OpenTelemetry::SpanHolder span("CPU_LEASE_DOWNSCALED");
+        span.addAttribute("workload", parent->settings.workload);
+        span.addAttribute("lease_id", parent->getLeaseId());
+        span.addAttribute("thread_number", slot_id);
+    }
+    parent.reset();
+}
+
+CPULeaseAllocation::CPULeaseAllocation(SlotCount max_threads_, ResourceLink master_link_, ResourceLink worker_link_, CPULeaseSettings settings_)
+    : max_threads(max_threads_)
+    , master_link(master_link_)
+    , worker_link(worker_link_)
+    , settings(std::move(settings_))
+    , log(getLogger("CPULeaseAllocation"))
+    , threads(max_threads)
+    , requests(max_threads) // NOTE: it should not be reallocated after initialization because we use raw pointers and iterators
+    , head(requests.begin())
+    , tail(head)
+    , acquired_increment(CurrentMetrics::ConcurrencyControlAcquired, 0)
+    , scheduled_increment(CurrentMetrics::ConcurrencyControlScheduled, 0)
+    , lease_id(lease_counter.fetch_add(1, std::memory_order_relaxed))
+{
+    chassert(max_threads > 0);
+    for (Request & request : requests)
+        request.lease = this;
+    std::unique_lock lock{mutex};
```

Comment:
Looks like a very complex synchronization. Why we need to lock something in constructor of class with mutex of this class.

Reply:
It's just an interface thing: `schedule()` requires a lock. I'm not sure how to improve it. I believe TSA annotations would also require a lock, even in ctor.